swap/src/bitcoin/wallet.rs

```diff
@@ -214,26 +215,21 @@ impl Wallet {
     }
 
     pub async fn subscribe_to(&self, tx: impl Watchable + Send + 'static) -> Subscription {
-        if let Some(sub) = self
-            .client
-            .lock()
-            .await
-            .subscriptions
-            .get(&WatchableTransaction {
-                txid: tx.id(),
-                script: tx.script(),
-            })
-        {
-            return Subscription {
-                receiver: sub.clone(),
-                finality_confirmations: self.finality_confirmations,
-                txid: tx.id(),
-            };
+        let (sender, receiver) = watch::channel(ScriptStatus::Unseen);
+        let txid = tx.id();
+
+        let sub = Subscription {
+            receiver,
+            finality_confirmations: self.finality_confirmations,
+            txid,
+            script: tx.script(),
         };
 
-        let (sender, receiver) = watch::channel(ScriptStatus::Unseen);
+        if let Some(inner) = self.client.lock().await.subscriptions.get(&sub) {
```

Comment:
It is kind of weird to construct `Subscription` just for the lookup and then potentially throw it away if you can find an exact copy in the HashSet. Can I suggest to use as `HashMap` instead where the key is either a distinct type or maybe just a tuple of TxId and Script? That will also get rid of the custom implementation of `PartialEq` and `Hash` which are not ideal given that this type is publicly exposed from this module.

Also, because we are locking here twice (once here and once on line 261), this is actually subject to a race condition.

We should do the locking only once to make sure this can't happen. You might also want to look into the `Entry` API of `HashMap`. That one allows you to ergonomically handle a "return or insert if not present" case.